File: backend-api/src/main/java/com/icefaces/modernization/api/service/CityService.java
package com.icefaces.modernization.api.service;

import com.icefaces.modernization.api.dto.CityDto;
import com.icefaces.modernization.api.dto.CityStatisticsDto;
import java.util.List;
import java.util.Optional;

/**
 * Service interface for City operations
 * Replaces business logic from ICEfaces managed beans like:
 * - org.icefaces.samples.showcase.example.ace.autocompleteentry.AutoCompleteEntryBean
 * - org.icefaces.samples.showcase.example.compat.autocomplete.AutocompleteBean
 * - Geographic/map-related managed beans
 * 
 * This interface defines the contract for city-related operations,
 * particularly focused on autocomplete and geographic functionality.
 */
public interface CityService {
    
    /**
     * Retrieve all cities
     * Replaces: ICEfaces managed bean city list initialization
     */
    List<CityDto> getAllCities();
    
    /**
     * Find city by ID
     * Replaces: ICEfaces managed bean selection logic
     */
    Optional<CityDto> getCityById(Long id);
    
    /**
     * Create new city
     * Replaces: ICEfaces managed bean add city functionality
     */
    CityDto createCity(CityDto cityDto);
    
    /**
     * Update existing city
     * Replaces: ICEfaces managed bean update city functionality
     */
    CityDto updateCity(Long id, CityDto cityDto);
    
    /**
     * Delete city
     * Replaces: ICEfaces managed bean delete city functionality
     */
    void deleteCity(Long id);
    
    /**
     * Autocomplete cities by name (optimized for fast response)
     * Replaces: ICEfaces autocomplete component data provision
     * This is the primary method for autocomplete functionality
     */
    List<CityDto> autocompleteCities(String searchTerm);
    
    /**
     * Search cities by name containing term
     * Replaces: ICEfaces search/filter functionality
     */
    List<CityDto> searchCitiesByName(String name);
    
    /**
     * Find cities by country
     * Replaces: ICEfaces country-based filtering
     */
    List<CityDto> getCitiesByCountry(String country);
    
    /**
     * Find exact city by name and country
     * Replaces: ICEfaces precise city lookup logic
     */
    Optional<CityDto> findCityByNameAndCountry(String name, String country);
    
    /**
     * Find cities within geographic bounding box
     * Replaces: ICEfaces map component data filtering
     */
    List<CityDto> getCitiesInBoundingBox(Double minLat, Double maxLat, 
                                        Double minLon, Double maxLon);
    
    /**
     * Find cities near a point (simplified distance)
     * Replaces: ICEfaces proximity search functionality
     */
    List<CityDto> getCitiesNearPoint(Double latitude, Double longitude, Double radius);
    
    /**
     * Get all countries for dropdown population
     * Replaces: ICEfaces managed bean methods for populating country selects
     */
    List<String> getAllCountries();
    
    /**
     * Get cities by country with ordering
     * Replaces: ICEfaces ordered city list generation
     */
    List<CityDto> getCitiesByCountryOrdered(String country);
    
    /**
     * Advanced city search with multiple criteria
     * Replaces: Complex ICEfaces search combinations
     */
    List<CityDto> searchCities(String name, String country);
    
    /**
     * Get city statistics for reporting
     * Replaces: ICEfaces statistical data generation
     */
    CityStatisticsDto getCityStatistics();
}

File: frontend-app/src/app/services/city.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface City {
  id?: number;
  name: string;
  country: string;
  latitude: number;
  longitude: number;
  altitude?: number;
  formattedCoordinates?: string;
  formattedAltitude?: string;
  autocompleteLabel?: string;
  autocompleteValue?: string;
  displayName?: string;
}

export interface CityStatistics {
  totalCities: number;
  totalCountries: number;
  averageLatitude: number;
  averageLongitude: number;
  averageAltitude: number;
}

@Injectable({
  providedIn: 'root'
})
export class CityService {
  private baseUrl = 'http://localhost:8081/api/v1/cities';

  constructor(private http: HttpClient) {}

  getAllCities(): Observable<City[]> {
    return this.http.get<City[]>(this.baseUrl);
  }

  getCityById(id: number): Observable<City> {
    return this.http.get<City>(`${this.baseUrl}/${id}`);
  }

  createCity(city: City): Observable<City> {
    return this.http.post<City>(this.baseUrl, city);
  }

  updateCity(id: number, city: City): Observable<City> {
    return this.http.put<City>(`${this.baseUrl}/${id}`, city);
  }

  deleteCity(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }

  autocompleteCities(query: string): Observable<City[]> {
    const params = new HttpParams().set('query', query);
    return this.http.get<City[]>(`${this.baseUrl}/autocomplete`, { params });
  }

  searchCitiesByName(name: string): Observable<City[]> {
    const params = new HttpParams().set('name', name);
    return this.http.get<City[]>(`${this.baseUrl}/search/name`, { params });
  }

  getCitiesByCountry(country: string): Observable<City[]> {
    return this.http.get<City[]>(`${this.baseUrl}/country/${country}`);
  }

  getCitiesByCountryOrdered(country: string): Observable<City[]> {
    return this.http.get<City[]>(`${this.baseUrl}/country/${country}/ordered`);
  }

  findCityByNameAndCountry(name: string, country: string): Observable<City> {
    const params = new HttpParams()
      .set('name', name)
      .set('country', country);
    return this.http.get<City>(`${this.baseUrl}/find`, { params });
  }

  getCitiesInBoundingBox(minLat: number, maxLat: number, minLon: number, maxLon: number): Observable<City[]> {
    const params = new HttpParams()
      .set('minLat', minLat.toString())
      .set('maxLat', maxLat.toString())
      .set('minLon', minLon.toString())
      .set('maxLon', maxLon.toString());
    return this.http.get<City[]>(`${this.baseUrl}/geo/bounding-box`, { params });
  }

  getCitiesNearPoint(latitude: number, longitude: number, radius: number): Observable<City[]> {
    const params = new HttpParams()
      .set('latitude', latitude.toString())
      .set('longitude', longitude.toString())
      .set('radius', radius.toString());
    return this.http.get<City[]>(`${this.baseUrl}/geo/near`, { params });
  }

  getAllCountries(): Observable<string[]> {
    return this.http.get<string[]>(`${this.baseUrl}/countries`);
  }

  searchCitiesAdvanced(name?: string, country?: string): Observable<City[]> {
    let params = new HttpParams();
    if (name) params = params.set('name', name);
    if (country) params = params.set('country', country);
    
    return this.http.get<City[]>(`${this.baseUrl}/search/advanced`, { params });
  }

  getCityStatistics(): Observable<CityStatistics> {
    return this.http.get<CityStatistics>(`${this.baseUrl}/statistics`);
  }
}